### PR TITLE
docs: update spec for Row Variables and no-input-extensions

### LINKED ISSUE #745: Rewrite extension section of spec to reflect current behaviour
And reword/use notation consistent with rest of spec.

### LINKED ISSUE #1097: Update spec wrt RowVariables

### PR DESCRIPTION
**Pull request by acl-cqc**
Also some driveby updates e.g. missed in #906.
Closes #1097, #745.

### CODE REVIEW

File: specification/hugr.md [hunk 1]
```diff
@@ -791,15 +791,16 @@ There are two classes of type: `AnyType` $\supset$ `CopyableType`. Types in thes
 classes are distinguished by whether the runtime values of those types can be implicitly
 copied or discarded (multiple or 0 links from on output port respectively):
 
-- For the broadest class (`AnyType`), the only operation supported is the identity operation (aka no-op, or `lift` - see [Extension Tracking](#extension-tracking) below). Specifically, we do not require it to be possible to copy or discard all values, hence the requirement that outports of linear type must have exactly one edge. (That is, a type not known to be in the copyable subset). All incoming ports must have exactly one edge.
+- For the broadest class (`AnyType`), the only operation supported is the identity operation (aka no-op, or `lift` - see [Extension Tracking](#extension-tracking) below). Specifically, we do not require it to be possible to copy or discard all values, hence the requirement that outports of linear type must have exactly one edge. (That is, a type not known to be in the copyable subset).
```
Comment (acl-cqc) [inline specification/hugr.md:794]: This was a driveby commoning up, see below

File: specification/hugr.md [hunk 2]
```diff
@@ -812,7 +813,7 @@ Extensions ::= (Extension)* -- a set, not a list
 Type ::= Sum([#]) -- disjoint union of rows of other types, tagged by unsigned int
        | Opaque(Name, [TypeArg]) -- a (instantiation of a) custom type defined by an extension
        | Function(#, #, Extensions) -- monomorphic function: arguments, results, and delta (see below)
-       | Variable -- refers to a TypeParam bound by the nearest enclosing FuncDefn node, or an enclosing Function Type
+       | Variable -- refers to a TypeParam bound by an the nearest enclosing FuncDefn node or polymorphic type scheme
```
Comment (acl-cqc) [inline specification/hugr.md:816]: This dates back to #906!

**Comment by ss2165** [comment]
I don't think spec should be explicit about inference, I think that is more a feature of the reference implementation.

File: specification/hugr.md [hunk 3]
```diff
@@ -868,12 +869,39 @@ a valid type as long as the TypeArgs match the declared TypeParams, which can be
...
+
 ### Extension Tracking
 
 The type of `Function` includes a set of [extensions](#extension-system) which are required to execute the graph.
-Every node in the HUGR is annotated with the set of extensions required to produce its inputs,
-and each operation provides (a way to compute) the set of extensions required to execute the node, known as the "delta";
-the union of these two must match the set of extensions on each successor node.
+Similarly, every dataflow node in the HUGR has a set of extensions required to execute the node (computed from its operation), also known as the "delta". The delta of any node must be a subset of its parent,
+except for FuncDefn's:
+* the delta of any child of a FuncDefn, must be a subset of the extensions in the FuncDefn's *type*
```
Comment (acl-cqc) [inline specification/hugr.md:901]: This is a bit awkward. Suggestions welcome

File: specification/hugr.md [hunk 4]
```diff
@@ -890,36 +918,24 @@ running different extensions. By the same mechanism, the runtime can reason
...
-X to the requirements of the
-graph.
+The latter is useful for higher-order operations such as conditionally selecting
+one function or another, where the output must have a consistent type (including
+the extension-requirements of the function).
 
-Having these as explicit nodes on the graph allows us to search for the
-point before extensions were added when we want to copy graphs, allowing
-us to get the version with minimal extension requirements.
+# Having these as explicit nodes on the graph allows us to search for the
```
Comment (acl-cqc) [inline specification/hugr.md:931]: Any point in keeping these sections I've commented out?

File: specification/hugr.md [hunk 5]
```diff
@@ -1147,6 +1163,14 @@ extensions:
...
     # - but default type scheme interpreter does not support such addition
     # Hence, no signature block => will look up a compute_signature in registry.
+  - name: TupleConcat
+    description: "Concatenate two tuples"
+    params:
+      a: List[Type]
+      b: List[Type]
+    signature:
+      inputs: Sum([a]), Sum([b])
+      outputs: Sum([(a,b)]) # Syntax! Sum([a,b]) would be either-a-or-b, this has one variant
```
Comment (acl-cqc) [inline specification/hugr.md:1173]: Bleurgh. It'd be awkward to write these examples *not* in the "declarative format", but OTOH I don't think we are really moving in this direction, and it's very unclear how to write these ATM.

File: specification/hugr.md [hunk 6]
```diff
@@ -868,12 +869,39 @@ a valid type as long as the TypeArgs match the declared TypeParams, which can be
...
+
+Note that since a row variable does not have kind Type, it cannot be used as the type of an edge.
+
 ### Extension Tracking
 
 The type of `Function` includes a set of [extensions](#extension-system) which are required to execute the graph.
-Every node in the HUGR is annotated with the set of extensions required to produce its inputs,
-and each operation provides (a way to compute) the set of extensions required to execute the node, known as the "delta";
-the union of these two must match the set of extensions on each successor node.
+Similarly, every dataflow node in the HUGR has a set of extensions required to execute the node (computed from its operation), also known as the "delta". The delta of any node must be a subset of its parent,
```
Comment (ss2165) [inline specification/hugr.md:899]:
```suggestion
Similarly, every dataflow node in the HUGR has a set of extensions required to execute the node (computed from its operation), also known as the "delta". The delta of any node must be a subset of its parent's delta,
```

File: specification/hugr.md (hunk 3 above)
Comment (ss2165) [reply specification/hugr.md:901]: just remove the comma I think

File: specification/hugr.md (hunk 4 above)
Comment (ss2165) [reply specification/hugr.md:931]: don't think so

File: specification/hugr.md (hunk 5 above)
Comment (ss2165) [reply specification/hugr.md:1173]: I think better just to say what it is in the comment
```suggestion
      outputs: Sum([(a,b)]) # sum with a single variant of type (a, b)
```

**Review by ss2165 — APPROVED** [review]
thanks!

File: specification/hugr.md (hunk 2 above)
Comment (acl-cqc) [reply specification/hugr.md:816]: Oops, `by an the` back to `by an`
